<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />


<meta name="author" content="Yusha Liu" />


<title>Application of poisson mash with random effects to single cell cytokines data (updated)</title>

<script src="site_libs/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<script src="site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>
<script src="site_libs/jqueryui-1.11.4/jquery-ui.min.js"></script>
<link href="site_libs/tocify-1.9.1/jquery.tocify.css" rel="stylesheet" />
<script src="site_libs/tocify-1.9.1/jquery.tocify.js"></script>
<script src="site_libs/navigation-1.1/tabsets.js"></script>
<link href="site_libs/highlightjs-9.12.0/textmate.css" rel="stylesheet" />
<script src="site_libs/highlightjs-9.12.0/highlight.js"></script>
<link href="site_libs/font-awesome-5.1.0/css/all.css" rel="stylesheet" />
<link href="site_libs/font-awesome-5.1.0/css/v4-shims.css" rel="stylesheet" />

<link rel="icon" href="https://github.com/workflowr/workflowr-assets/raw/master/img/reproducible.png">
<!-- Add a small amount of space between sections. -->
<style type="text/css">
div.section {
  padding-top: 12px;
}
</style>

<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs) {
  hljs.configure({languages: []});
  hljs.initHighlightingOnLoad();
  if (document.readyState && document.readyState === "complete") {
    window.setTimeout(function() { hljs.initHighlighting(); }, 0);
  }
}
</script>



<style type="text/css">
h1 {
  font-size: 34px;
}
h1.title {
  font-size: 38px;
}
h2 {
  font-size: 30px;
}
h3 {
  font-size: 24px;
}
h4 {
  font-size: 18px;
}
h5 {
  font-size: 16px;
}
h6 {
  font-size: 12px;
}
.table th:not([align]) {
  text-align: left;
}
</style>




<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img {
  max-width:100%;
}
.tabbed-pane {
  padding-top: 12px;
}
.html-widget {
  margin-bottom: 20px;
}
button.code-folding-btn:focus {
  outline: none;
}
summary {
  display: list-item;
}
</style>


<style type="text/css">
/* padding for bootstrap navbar */
body {
  padding-top: 51px;
  padding-bottom: 40px;
}
/* offset scroll position for anchor links (for fixed navbar)  */
.section h1 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h2 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h3 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h4 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h5 {
  padding-top: 56px;
  margin-top: -56px;
}
.section h6 {
  padding-top: 56px;
  margin-top: -56px;
}
.dropdown-submenu {
  position: relative;
}
.dropdown-submenu>.dropdown-menu {
  top: 0;
  left: 100%;
  margin-top: -6px;
  margin-left: -1px;
  border-radius: 0 6px 6px 6px;
}
.dropdown-submenu:hover>.dropdown-menu {
  display: block;
}
.dropdown-submenu>a:after {
  display: block;
  content: " ";
  float: right;
  width: 0;
  height: 0;
  border-color: transparent;
  border-style: solid;
  border-width: 5px 0 5px 5px;
  border-left-color: #cccccc;
  margin-top: 5px;
  margin-right: -10px;
}
.dropdown-submenu:hover>a:after {
  border-left-color: #ffffff;
}
.dropdown-submenu.pull-left {
  float: none;
}
.dropdown-submenu.pull-left>.dropdown-menu {
  left: -100%;
  margin-left: 10px;
  border-radius: 6px 0 6px 6px;
}
</style>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');

  // mark it active
  menuAnchor.parent().addClass('active');

  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>

<!-- tabsets -->

<style type="text/css">
.tabset-dropdown > .nav-tabs {
  display: inline-table;
  max-height: 500px;
  min-height: 44px;
  overflow-y: auto;
  background: white;
  border: 1px solid #ddd;
  border-radius: 4px;
}

.tabset-dropdown > .nav-tabs > li.active:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li.active:before {
  content: "&#xe258;";
  border: none;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open:before {
  content: "";
  font-family: 'Glyphicons Halflings';
  display: inline-block;
  padding: 10px;
  border-right: 1px solid #ddd;
}

.tabset-dropdown > .nav-tabs > li.active {
  display: block;
}

.tabset-dropdown > .nav-tabs > li > a,
.tabset-dropdown > .nav-tabs > li > a:focus,
.tabset-dropdown > .nav-tabs > li > a:hover {
  border: none;
  display: inline-block;
  border-radius: 4px;
  background-color: transparent;
}

.tabset-dropdown > .nav-tabs.nav-tabs-open > li {
  display: block;
  float: none;
}

.tabset-dropdown > .nav-tabs > li {
  display: none;
}
</style>

<!-- code folding -->



<style type="text/css">

#TOC {
  margin: 25px 0px 20px 0px;
}
@media (max-width: 768px) {
#TOC {
  position: relative;
  width: 100%;
}
}

@media print {
.toc-content {
  /* see https://github.com/w3c/csswg-drafts/issues/4434 */
  float: right;
}
}

.toc-content {
  padding-left: 30px;
  padding-right: 40px;
}

div.main-container {
  max-width: 1200px;
}

div.tocify {
  width: 20%;
  max-width: 260px;
  max-height: 85%;
}

@media (min-width: 768px) and (max-width: 991px) {
  div.tocify {
    width: 25%;
  }
}

@media (max-width: 767px) {
  div.tocify {
    width: 100%;
    max-width: none;
  }
}

.tocify ul, .tocify li {
  line-height: 20px;
}

.tocify-subheader .tocify-item {
  font-size: 0.90em;
}

.tocify .list-group-item {
  border-radius: 0px;
}


</style>



</head>

<body>


<div class="container-fluid main-container">


<!-- setup 3col/9col grid for toc_float and main content  -->
<div class="row-fluid">
<div class="col-xs-12 col-sm-4 col-md-3">
<div id="TOC" class="tocify">
</div>
</div>

<div class="toc-content col-xs-12 col-sm-8 col-md-9">




<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="index.html">poisson_mash</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        <li>
  <a href="index.html">Home</a>
</li>
<li>
  <a href="about.html">About</a>
</li>
<li>
  <a href="license.html">License</a>
</li>
      </ul>
      <ul class="nav navbar-nav navbar-right">
        <li>
  <a href="https://github.com/YushaLiu/poisson_mash">
    <span class="fa fa-github"></span>
     
    Source code
  </a>
</li>
      </ul>
    </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

<div class="fluid-row" id="header">



<h1 class="title toc-ignore">Application of poisson mash with random effects to single cell cytokines data (updated)</h1>
<h4 class="author">Yusha Liu</h4>
<h4 class="date">2020-8-19</h4>

</div>


<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-report" data-toggle="collapse" data-target="#workflowr-report">
<span class="glyphicon glyphicon-list" aria-hidden="true"></span> workflowr <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span>
</button>
</p>
<div id="workflowr-report" class="collapse">
<ul class="nav nav-tabs">
<li class="active">
<a data-toggle="tab" href="#summary">Summary</a>
</li>
<li>
<a data-toggle="tab" href="#checks"> Checks <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> </a>
</li>
<li>
<a data-toggle="tab" href="#versions">Past versions</a>
</li>
</ul>
<div class="tab-content">
<div id="summary" class="tab-pane fade in active">
<p>
<strong>Last updated:</strong> 2020-08-20
</p>
<p>
<strong>Checks:</strong> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> 7 <span class="glyphicon glyphicon-exclamation-sign text-danger" aria-hidden="true"></span> 0
</p>
<p>
<strong>Knit directory:</strong> <code>poisson_mash/</code> <span class="glyphicon glyphicon-question-sign" aria-hidden="true" title="This is the local directory in which the code in this file was executed."> </span>
</p>
<p>
This reproducible <a href="http://rmarkdown.rstudio.com">R Markdown</a> analysis was created with <a
  href="https://github.com/jdblischak/workflowr">workflowr</a> (version 1.6.2). The <em>Checks</em> tab describes the reproducibility checks that were applied when the results were created. The <em>Past versions</em> tab lists the development history.
</p>
<hr>
</div>
<div id="checks" class="tab-pane fade">
<div id="workflowr-checks" class="panel-group">
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRMarkdownfilestronguptodate"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>R Markdown file:</strong> up-to-date </a>
</p>
</div>
<div id="strongRMarkdownfilestronguptodate" class="panel-collapse collapse">
<div class="panel-body">
<p>Great! Since the R Markdown file has been committed to the Git repository, you know the exact version of the code that produced these results.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongEnvironmentstrongempty"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Environment:</strong> empty </a>
</p>
</div>
<div id="strongEnvironmentstrongempty" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! The global environment was empty. Objects defined in the global environment can affect the analysis in your R Markdown file in unknown ways. For reproduciblity it’s best to always run the code in an empty environment.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSeedstrongcodesetseed20200618code"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Seed:</strong> <code>set.seed(20200618)</code> </a>
</p>
</div>
<div id="strongSeedstrongcodesetseed20200618code" class="panel-collapse collapse">
<div class="panel-body">
<p>The command <code>set.seed(20200618)</code> was run prior to running the code in the R Markdown file. Setting a seed ensures that any results that rely on randomness, e.g. subsampling or permutations, are reproducible.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongSessioninformationstrongrecorded"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Session information:</strong> recorded </a>
</p>
</div>
<div id="strongSessioninformationstrongrecorded" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Recording the operating system, R version, and package versions is critical for reproducibility.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongCachestrongnone"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Cache:</strong> none </a>
</p>
</div>
<div id="strongCachestrongnone" class="panel-collapse collapse">
<div class="panel-body">
<p>Nice! There were no cached chunks for this analysis, so you can be confident that you successfully produced the results during this run.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongFilepathsstrongrelative"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>File paths:</strong> relative </a>
</p>
</div>
<div id="strongFilepathsstrongrelative" class="panel-collapse collapse">
<div class="panel-body">
<p>Great job! Using relative paths to the files within your workflowr project makes it easier to run your code on other machines.</p>
</div>
</div>
</div>
<div class="panel panel-default">
<div class="panel-heading">
<p class="panel-title">
<a data-toggle="collapse" data-parent="#workflowr-checks" href="#strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashtree2bf72bbbf6490c2fc9d712096c1cc40d7e4f5258targetblank2bf72bba"> <span class="glyphicon glyphicon-ok text-success" aria-hidden="true"></span> <strong>Repository version:</strong> <a href="https://github.com/YushaLiu/poisson_mash/tree/2bf72bbbf6490c2fc9d712096c1cc40d7e4f5258" target="_blank">2bf72bb</a> </a>
</p>
</div>
<div id="strongRepositoryversionstrongahrefhttpsgithubcomYushaLiupoissonmashtree2bf72bbbf6490c2fc9d712096c1cc40d7e4f5258targetblank2bf72bba" class="panel-collapse collapse">
<div class="panel-body">
<p>
Great! You are using Git for version control. Tracking code development and connecting the code version to the results is critical for reproducibility.
</p>
<p>
The results in this page were generated with repository version <a href="https://github.com/YushaLiu/poisson_mash/tree/2bf72bbbf6490c2fc9d712096c1cc40d7e4f5258" target="_blank">2bf72bb</a>. See the <em>Past versions</em> tab to see a history of the changes made to the R Markdown and HTML files.
</p>
<p>
Note that you need to be careful to ensure that all relevant files for the analysis have been committed to Git prior to generating the results (you can use <code>wflow_publish</code> or <code>wflow_git_commit</code>). workflowr only checks the R Markdown file, but you know if there are other scripts or data files that it depends on. Below is the status of the Git repository when the results were generated:
</p>
<pre><code>
Ignored files:
    Ignored:    .DS_Store
    Ignored:    .Rhistory
    Ignored:    code/.DS_Store
    Ignored:    output/.DS_Store

Untracked files:
    Untracked:  code/code_random_effects.R
    Untracked:  code/examples.R
    Untracked:  code/examples_updated.R
    Untracked:  code/util6.R
    Untracked:  code/util8.R
    Untracked:  code/util_ed.R
    Untracked:  code/util_random_effects.R
    Untracked:  code/util_random_effects_v2.R
    Untracked:  data/data_jr.RData
    Untracked:  output/ed_all_trts_maxiter1500.Rds
    Untracked:  output/ed_second_batch_maxiter1500.Rds
    Untracked:  output/ed_second_batch_v2_maxiter2000.Rds
    Untracked:  output/ed_v1_maxiter1000.Rds
    Untracked:  output/ed_v2_maxiter1000.Rds
    Untracked:  output/examples.pdf
    Untracked:  output/examples/
    Untracked:  output/examples_mash.pdf
    Untracked:  output/mash_data_limma.rds
    Untracked:  output/mash_data_second_batch.rds
    Untracked:  output/mash_fit_limma.rds
    Untracked:  output/mash_fit_limma_all_trts.rds
    Untracked:  output/mash_fit_second_batch.rds
    Untracked:  output/non_null_sim_setting1/
    Untracked:  output/non_null_sim_setting2/
    Untracked:  output/non_null_sim_setting3/
    Untracked:  output/non_null_sim_setting4/
    Untracked:  output/null_sim/
    Untracked:  output/pois_mash_contrast_posterior_all_trts.Rds
    Untracked:  output/pois_mash_contrast_posterior_second_batch.Rds
    Untracked:  output/pois_mash_contrast_posterior_second_batch_v2.Rds
    Untracked:  output/pois_mash_fit_all_trts.Rds
    Untracked:  output/pois_mash_fit_second_batch.Rds
    Untracked:  output/pois_mash_fit_second_batch_v2.Rds
    Untracked:  output/poisson_mash_fit.Rds
    Untracked:  output/poisson_mash_fit_contrast_posterior.Rds
    Untracked:  output/realdata/
    Untracked:  output/simulations/

</code></pre>
<p>
Note that any generated files, e.g. HTML, png, CSS, etc., are not included in this status report because it is ok for generated content to have uncommitted changes.
</p>
</div>
</div>
</div>
</div>
<hr>
</div>
<div id="versions" class="tab-pane fade">

<p>
These are the previous versions of the repository in which changes were made to the R Markdown (<code>analysis/pois_mash_random_effects_updated.Rmd</code>) and HTML (<code>docs/pois_mash_random_effects_updated.html</code>) files. If you’ve configured a remote Git repository (see <code>?wflow_git_remote</code>), click on the hyperlinks in the table below to view the files as they were in that past version.
</p>
<div class="table-responsive">
<table class="table table-condensed table-hover">
<thead>
<tr>
<th>
File
</th>
<th>
Version
</th>
<th>
Author
</th>
<th>
Date
</th>
<th>
Message
</th>
</tr>
</thead>
<tbody>
<tr>
<td>
Rmd
</td>
<td>
<a href="https://github.com/YushaLiu/poisson_mash/blob/2bf72bbbf6490c2fc9d712096c1cc40d7e4f5258/analysis/pois_mash_random_effects_updated.Rmd" target="_blank">2bf72bb</a>
</td>
<td>
yushaliu
</td>
<td>
2020-08-20
</td>
<td>
Add more results
</td>
</tr>
</tbody>
</table>
</div>
<hr>
</div>
</div>
</div>
<div id="summary" class="section level2">
<h2>Summary</h2>
<p>We apply poisson mash with random effects to CD4 T cells of the single cell cytokines data. The four conditions from the first batch (Ctrl_1, IFNa, IFNb, TNF) are removed from analysis for now. The count data matrix <span class="math inline">\(X\)</span> is <span class="math inline">\(14,853 \times 46\)</span>, where <span class="math inline">\(X_{jr}\)</span> represents the total read counts of gene <span class="math inline">\(j\)</span> in all cells from condition <span class="math inline">\(r\)</span>.</p>
<pre class="r"><code>load(&quot;data/data_jr.RData&quot;)
source(&quot;code/code_random_effects.R&quot;)

data &lt;- data.jr
s &lt;- s.r/8
rm(data.jr, s.r)

data &lt;- data[,5:50]
s &lt;- s[5:50]

dim(data)</code></pre>
<pre><code>[1] 14853    46</code></pre>
<div id="model" class="section level3">
<h3>Model</h3>
<p>We assume <span class="math inline">\(X_{jr} \sim Pois(s_r \lambda_{jr})\)</span>, and <span class="math inline">\(\log(\lambda_{jr}) = \mu_j + \beta_{jr} + \eta_{jr}\)</span>, where <span class="math inline">\(s_r\)</span> accounts for the different sequencing depth of each condition, <span class="math inline">\(\mu_j\)</span> is the gene-specific mean on the log scale shared by all conditions, <span class="math inline">\(\beta_{jr}\)</span> models the condition-specific treatment effect, and <span class="math inline">\(\eta_{jr}\)</span> models the random effect. We are mainly interested in estimation and inference on <span class="math inline">\(\beta_{jr}\)</span>. The following priors are assumed in our model: <span class="math display">\[\begin{align}
&amp; \beta_j  =  \sum_{h, l} \pi_{h, l} \: MVN(0, \: w_l U_h) + \sum_{g, l} \pi_{g, l} \: MVN(0, \: w_l u_g u&#39;_g) \\
\text{where} \quad &amp; \sum_{h, l} \pi_{h, l} + \sum_{g, l} \pi_{g, l} = 1 \\
&amp; \eta_j \sim MVN(0, \: \tau_j^2 I_R). \\
\end{align}\]</span></p>
</div>
<div id="ed-step" class="section level3">
<h3>ED step</h3>
<p>To estimate the data-driven prior covariances <span class="math inline">\(U_h\)</span> and <span class="math inline">\(u_g u&#39;_g\)</span>, we fit the poisson mash model without scaling parameters <span class="math inline">\(w_l\)</span> to a subset of genes that are differentially expressed across conditions, which are identified based on a conditional multinomial goodness-of-fit test. To capture as many effect-sharing patterns as possible, we choose to set the number of principal components in the ED step to be large.</p>
<pre class="r"><code>##  identify the subset of differentially expressed genes and initialize the data-driven prior covariances
res.pca &lt;- pois_cov_init(data, s, npc=26)
subset &lt;- res.pca$subset
data.ed &lt;- data[subset,]

## the number of differentially expressed genes identified by the multinomial goodness-of-fit test
dim(data.ed)</code></pre>
<pre><code>[1] 2481   46</code></pre>
<pre class="r"><code>## load in the ED results
fit.ed &lt;- readRDS(&quot;output/realdata/pois_mash_ed_maxiter2000.Rds&quot;)

## running time per iteration in ED
fit.ed$runtime/length(fit.ed$ELBO)</code></pre>
<pre><code>     user    system   elapsed 
37.879193  0.058182 37.951517 </code></pre>
<pre class="r"><code>## estimated weights for different candidate data-driven prior covariances
names(fit.ed$pi) &lt;- c(names(res.pca$Ulist), names(res.pca$ulist), &quot;u_0&quot;)
fit.ed$pi</code></pre>
<pre><code>       tPCA     Emp_cov        PC_1        PC_2        PC_3        PC_4 
0.010433158 0.001241230 0.100145969 0.052509367 0.058669542 0.076081041 
       PC_5        PC_6        PC_7        PC_8        PC_9       PC_10 
0.043761883 0.011665023 0.047405971 0.020133862 0.009708747 0.006999201 
      PC_11       PC_12       PC_13       PC_14       PC_15       PC_16 
0.050142317 0.027939154 0.038560296 0.008186151 0.099139253 0.040466414 
      PC_17       PC_18       PC_19       PC_20       PC_21       PC_22 
0.002053865 0.072518077 0.003674665 0.001178986 0.001849840 0.002125615 
      PC_23       PC_24       PC_25       PC_26         u_0 
0.102451143 0.057934033 0.044352513 0.008671866 0.000001000 </code></pre>
<pre class="r"><code>## look at the ELBO
plot(fit.ed$ELBO, xlab = &quot;niter&quot;, ylab = &quot;elbo&quot;, main=&quot;fitting ED step&quot;)</code></pre>
<p><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-2-1.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="estimate-of-data-driven-covariances-from-ed-step" class="section level3">
<h3>Estimate of data-driven covariances from ED step</h3>
<pre class="r"><code>library(RColorBrewer)

cols.all &lt;- c(&quot;blue&quot;, brewer.pal(n=5,name=&quot;Greens&quot;)[3:5], &quot;gray&quot;, brewer.pal(n=9,name=&quot;PuRd&quot;), 
              brewer.pal(n=9,name=&quot;Reds&quot;), brewer.pal(n=9,name=&quot;Oranges&quot;),
              brewer.pal(n=9,name=&quot;Purples&quot;), brewer.pal(n=9,name=&quot;YlOrRd&quot;))

cols.trt &lt;- cols.all[5:50]

par(mfrow=c(2,2))

pi.g &lt;- fit.ed$pi[-c(1:2)]
ulist &lt;- fit.ed$ulist
names(ulist) &lt;- names(fit.ed$pi[-c(1:2)])

### plot the data-driven u_g&#39;s with estimated weights larger than or equal to 2%
ulist &lt;- ulist[pi.g &gt;= 2e-2]
pi.g &lt;- pi.g[pi.g &gt;= 2e-2]

for (k in 1:length(ulist)){
  v &lt;- ulist[[k]]
  
  barplot(v/v[which.max(abs(v))], names = names(s), cex.names = 0.4, las = 2, 
          main = paste0(names(ulist)[k], &quot;: weight &quot;, round(pi.g[k],3)), col = cols.trt)
  
}</code></pre>
<p><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-3-1.png" width="672" style="display: block; margin: auto;" /><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-3-2.png" width="672" style="display: block; margin: auto;" /><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-3-3.png" width="672" style="display: block; margin: auto;" /><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-3-4.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="running-poisson-mash" class="section level3">
<h3>Running poisson mash</h3>
<p>We then fit the poisson mash model to all the genes in <span class="math inline">\(X\)</span> which have at least 20 total read counts in all conditions. There are 10,733 such genes.</p>
<pre class="r"><code>## remove genes that have fewer than 20 total read counts in all 46 conditions
data &lt;- data[rowSums(data) &gt;= 20,]
dim(data)</code></pre>
<pre><code>[1] 10733    46</code></pre>
<pre class="r"><code>## load in the poisson mash fit results
res &lt;- readRDS(&quot;output/realdata/pois_mash_fit.Rds&quot;)
res$runtime</code></pre>
<pre><code>     user    system   elapsed 
37090.804   335.217 37441.977 </code></pre>
<pre class="r"><code>## look at ELBO
plot(res$ELBO, xlab = &quot;niter&quot;, ylab = &quot;elbo&quot;, main=&quot;Fitting poisson mash with random effects&quot;)</code></pre>
<p><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-4-1.png" width="672" style="display: block; margin: auto;" /></p>
<pre class="r"><code>## look at the estimated weights of prior covariance matrices
library(pheatmap)
library(gridExtra)

wlist &lt;- c(1e-4, 4e-4, 1e-3, 4e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32, 0.64, 1, 2, 4, 8, 16)

pi.mat &lt;- matrix(res$pi, ncol=length(wlist), byrow=TRUE)

rownames(pi.mat) &lt;- c(&quot;tPCA&quot;, &quot;Emp_cov&quot;, paste0(&quot;e_&quot;, 1:46), paste0(&quot;PC_&quot;, 1:26), &quot;u_0&quot;)
colnames(pi.mat) &lt;- paste0(&quot;w=&quot;, wlist)
pheatmap(pi.mat, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=6, main=&quot;Estimated weights of prior covariances in poisson mash fit&quot;)</code></pre>
<p><img src="figure/pois_mash_random_effects_updated.Rmd/unnamed-chunk-4-2.png" width="672" style="display: block; margin: auto;" /></p>
</div>
<div id="comparison-to-fixed-effects-based-mash" class="section level3">
<h3>Comparison to fixed effects-based mash</h3>
<p>We calculate posterior summaries of <span class="math inline">\(\delta_{jr} = \beta_{jr} - \sum_{r=1}^R \beta_{jr}/R\)</span> from the poisson mash fit, i.e., the condition-specific deviation from the mean across all conditions, and compare to the results from fixed effects-based mash. In particular, we fit a linear fixed effects model separately to each gene <span class="math inline">\(j\)</span>, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the predictor, to obtain <span class="math inline">\(\hat{b}_{jr}\)</span> and <span class="math inline">\(\hat{s}_{jr}\)</span> for each gene <span class="math inline">\(j\)</span> and condition <span class="math inline">\(r\)</span>. When fitting this model, we assume that the Gaussian observation noise is constant over conditions for each gene <span class="math inline">\(j\)</span>. We then feed <span class="math inline">\(\hat{b}_{jr}\)</span> and <span class="math inline">\(\hat{s}_{jr}\)</span> to mash without common baseline. We consider only the 7,429 genes that have reads in at least <span class="math inline">\(46*5=230\)</span> cells.</p>
<pre class="r"><code>## load in the posterior summaries from poisson mash fit
post &lt;- readRDS(&quot;output/realdata/pois_mash_posterior.Rds&quot;)

## load in the mash fit based on bhat and shat obtained by fitting fixed effects model
mash.fixed &lt;- readRDS(&quot;output/realdata/mash_fit.rds&quot;)

## the number of significant genes identified by poisson mash (in 10,733 genes tested by poisson mash)
idx.pois.mash &lt;- which(apply(post$lfsr, 1, min) &lt; 0.05)
length(idx.pois.mash)</code></pre>
<pre><code>[1] 1341</code></pre>
<pre class="r"><code>## the number of significant genes identified by fixed effects-based mash (in 7,429 genes tested by mash)
idx.mash.fixed &lt;- mashr::get_significant_results(mash.fixed)
length(idx.mash.fixed)</code></pre>
<pre><code>[1] 768</code></pre>
<pre class="r"><code>## the number of genes that are identified by both poisson mash and fixed effects-based mash
sum(names(idx.mash.fixed) %in% names(idx.pois.mash))</code></pre>
<pre><code>[1] 604</code></pre>
<pre class="r"><code>## the number of genes that are identified only by poisson mash (in 7,429 genes tested by both)
pois_mash_only &lt;- idx.pois.mash[!(names(idx.pois.mash) %in% names(idx.mash.fixed)) &amp; (names(idx.pois.mash) %in% rownames(mash.fixed$result$lfsr))]
length(pois_mash_only)</code></pre>
<pre><code>[1] 612</code></pre>
<pre class="r"><code>## the number of genes that are identified only by mash 
mash_fixed_only &lt;- idx.mash.fixed[!(names(idx.mash.fixed) %in% names(idx.pois.mash))]
length(mash_fixed_only)</code></pre>
<pre><code>[1] 164</code></pre>
</div>
<div id="comparison-to-mixed-effects-based-mash" class="section level3">
<h3>Comparison to mixed effects-based mash</h3>
<p>We then compare results from the poisson mash to the results from mixed effects-based mash. In particular, we fit a linear mixed effects model separately to each gene <span class="math inline">\(j\)</span>, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the design matrix for both fixed and random effects, to obtain <span class="math inline">\(\hat{b}_{jr}\)</span> and <span class="math inline">\(\hat{s}_{jr}\)</span> for each gene <span class="math inline">\(j\)</span> and condition <span class="math inline">\(r\)</span>. We use the dream function in the R package variancePartition to fit this mixed effects model (<a href="https://www.bioconductor.org/packages/release/bioc/html/variancePartition.html" class="uri">https://www.bioconductor.org/packages/release/bioc/html/variancePartition.html</a>). We then feed <span class="math inline">\(\hat{b}_{jr}\)</span> and <span class="math inline">\(\hat{s}_{jr}\)</span> to mash without common baseline. Again, we consider only the 7,429 genes that have reads in at least <span class="math inline">\(46*5=230\)</span> cells.</p>
<pre class="r"><code>## load in the mash fit based on bhat and shat obtained by fitting mixed effects model
mash.mixed &lt;- readRDS(&quot;output/realdata/mash_dream_fit.Rds&quot;)

## the number of significant genes identified by mixed effects-based mash (in 7,429 genes tested by mash)
idx.mash.mixed &lt;- mashr::get_significant_results(mash.mixed)
length(idx.mash.mixed)</code></pre>
<pre><code>[1] 68</code></pre>
<pre class="r"><code>## the number of genes that are identified by both poisson mash and mixed effects-based mash
sum(names(idx.mash.mixed) %in% names(idx.pois.mash))</code></pre>
<pre><code>[1] 65</code></pre>
<br>
<p>
<button type="button" class="btn btn-default btn-workflowr btn-workflowr-sessioninfo" data-toggle="collapse" data-target="#workflowr-sessioninfo" style="display: block;">
<span class="glyphicon glyphicon-wrench" aria-hidden="true"></span> Session information
</button>
</p>
<div id="workflowr-sessioninfo" class="collapse">
<pre class="r"><code>sessionInfo()</code></pre>
<pre><code>R version 3.6.2 (2019-12-12)
Platform: x86_64-apple-darwin15.6.0 (64-bit)
Running under: macOS Catalina 10.15.6

Matrix products: default
BLAS:   /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRblas.0.dylib
LAPACK: /Library/Frameworks/R.framework/Versions/3.6/Resources/lib/libRlapack.dylib

locale:
[1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] gridExtra_2.3      pheatmap_1.0.12    RColorBrewer_1.1-2 rmeta_3.0         
[5] ggplot2_3.3.0      poilog_0.4         psych_1.9.12.31    flashr_0.6-7      
[9] workflowr_1.6.2   

loaded via a namespace (and not attached):
 [1] softImpute_1.4   tidyselect_1.1.0 xfun_0.13        ashr_2.2-50     
 [5] purrr_0.3.4      reshape2_1.4.4   lattice_0.20-41  colorspace_1.4-1
 [9] vctrs_0.3.0      htmltools_0.4.0  yaml_2.2.1       rlang_0.4.6     
[13] mixsqp_0.3-43    later_1.0.0      pillar_1.4.4     withr_2.2.0     
[17] glue_1.4.1       mashr_0.2.21     lifecycle_0.2.0  plyr_1.8.6      
[21] stringr_1.4.0    munsell_0.5.0    gtable_0.3.0     mvtnorm_1.1-0   
[25] evaluate_0.14    knitr_1.28       httpuv_1.5.2     invgamma_1.1    
[29] irlba_2.3.3      parallel_3.6.2   Rcpp_1.0.4.6     promises_1.1.0  
[33] scales_1.1.1     backports_1.1.7  abind_1.4-5      truncnorm_1.0-8 
[37] fs_1.4.1         mnormt_1.5-7     digest_0.6.25    stringi_1.4.6   
[41] dplyr_0.8.5      grid_3.6.2       rprojroot_1.3-2  tools_3.6.2     
[45] magrittr_1.5     tibble_3.0.1     crayon_1.3.4     whisker_0.4     
[49] pkgconfig_2.0.3  ellipsis_0.3.1   Matrix_1.2-18    SQUAREM_2020.2  
[53] assertthat_0.2.1 rmarkdown_2.1    R6_2.4.1         nlme_3.1-147    
[57] git2r_0.27.1     compiler_3.6.2  </code></pre>
</div>
</div>
</div>


<!-- Adjust MathJax settings so that all math formulae are shown using
TeX fonts only; see
http://docs.mathjax.org/en/latest/configuration.html.  This will make
the presentation more consistent at the cost of the webpage sometimes
taking slightly longer to load. Note that this only works because the
footer is added to webpages before the MathJax javascript. -->
<script type="text/x-mathjax-config">
  MathJax.Hub.Config({
    "HTML-CSS": { availableFonts: ["TeX"] }
  });
</script>


</div>
</div>

</div>

<script>

// add bootstrap table styles to pandoc tables
function bootstrapStylePandocTables() {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
}
$(document).ready(function () {
  bootstrapStylePandocTables();
});


</script>

<!-- tabsets -->

<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});

$(document).ready(function () {
  $('.tabset-dropdown > .nav-tabs > li').click(function () {
    $(this).parent().toggleClass('nav-tabs-open')
  });
});
</script>

<!-- code folding -->

<script>
$(document).ready(function ()  {

    // move toc-ignore selectors from section div to header
    $('div.section.toc-ignore')
        .removeClass('toc-ignore')
        .children('h1,h2,h3,h4,h5').addClass('toc-ignore');

    // establish options
    var options = {
      selectors: "h1,h2,h3",
      theme: "bootstrap3",
      context: '.toc-content',
      hashGenerator: function (text) {
        return text.replace(/[.\\/?&!#<>]/g, '').replace(/\s/g, '_').toLowerCase();
      },
      ignoreSelector: ".toc-ignore",
      scrollTo: 0
    };
    options.showAndHide = true;
    options.smoothScroll = true;

    // tocify
    var toc = $("#TOC").tocify(options).data("toc-tocify");
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
