---
title: "Examples of genes that are identified by mash with no common baseline only"
author: "Yusha Liu"
header-includes:
date: "2020-8-3"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Summary
We apply both poisson mash and mash to CD4 T cells of the single cell cytokines data. The four conditions from the first batch (Ctrl_1, IFNa, IFNb, TNF) are removed from analysis for now. 

```{r warning=FALSE, message=FALSE}
### load in functions
source("code/util_random_effects_v2.R")

### load in data
load("data/data_jr.RData")
data <- data.jr
s <- s.r/8
rm(data.jr, s.r)
data <- data[,5:50]
s <- s[5:50]

### remove genes that have fewer than 20 total counts across all 46 treatments
data <- data[rowSums(data) >= 20,]

### color code for treatments
library(RColorBrewer)
cols.all <- c("blue", brewer.pal(n=5,name="Greens")[3:5], "gray", brewer.pal(n=9,name="PuRd"), 
              brewer.pal(n=9,name="Reds"), brewer.pal(n=9,name="Oranges"),
              brewer.pal(n=9,name="Purples"), brewer.pal(n=9,name="YlOrRd"))
cols.trt <- cols.all[5:50]

## load in the ED results
fit.ed <- readRDS("output/ed_second_batch_v2_maxiter2000.Rds")

### load in posterior summary of pois mash
post <- readRDS("output/pois_mash_contrast_posterior_second_batch_v2.Rds")
idx.pois.mash <- which(apply(post$lfsr, 1, min) < 0.05)

### load in posterior summary of mash
mash.fit <- readRDS("output/mash_fit_second_batch.rds")
idx.mash <- mashr::get_significant_results(mash.fit)

## the number of genes that are identified only by mash 
mash_only <- idx.mash[!(names(idx.mash) %in% names(idx.pois.mash))]

### randomly select 10 genes to plot
set.seed(100)
genes.to.plot <- names(mash_only)[sample(1:length(mash_only), 10, replace = FALSE)]

```


Now we present some examples of genes that are identified to be differentially expressed across treatments only by mash without common baseline. 

```{r warning=FALSE, message=FALSE}
### a few examples
for(j in 1:10){
  create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[j], cols.trt=cols.trt)
}
```


