---
title: "Application of poisson mash with random effects to single cell cytokines data (updated)"
author: "Yusha Liu"
header-includes:
date: "2020-8-19"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Summary
We apply poisson mash with random effects to CD4 T cells of the single cell cytokines data. The four conditions from the first batch (Ctrl_1, IFNa, IFNb, TNF) are removed from analysis for now. The count data matrix $X$ is $14,853 \times 46$, where $X_{jr}$ represents the total read counts of gene $j$ in all cells from condition $r$. 
```{r warning=FALSE, message=FALSE}
load("data/data_jr.RData")
source("code/code_random_effects.R")

data <- data.jr
s <- s.r/8
rm(data.jr, s.r)

data <- data[,5:50]
s <- s[5:50]

dim(data)

```



### Model
We assume $X_{jr} \sim Pois(s_r \lambda_{jr})$, and $\log(\lambda_{jr}) = \mu_j + \beta_{jr} + \eta_{jr}$, where $s_r$ accounts for the different sequencing depth of each condition, $\mu_j$ is the gene-specific mean on the log scale shared by all conditions, $\beta_{jr}$ models the condition-specific treatment effect, and $\eta_{jr}$ models the random effect. We are mainly interested in estimation and inference on $\beta_{jr}$. The following priors are assumed in our model:
\begin{align}
& \beta_j  =  \sum_{h, l} \pi_{h, l} \: MVN(0, \: w_l U_h) + \sum_{g, l} \pi_{g, l} \: MVN(0, \: w_l u_g u'_g) \\
\text{where} \quad & \sum_{h, l} \pi_{h, l} + \sum_{g, l} \pi_{g, l} = 1 \\
& \eta_j \sim MVN(0, \: \tau_j^2 I_R). \\
\end{align}



### ED step
To estimate the data-driven prior covariances $U_h$ and $u_g u'_g$, we fit the poisson mash model without scaling parameters $w_l$ to a subset of genes that are differentially expressed across conditions, which are identified based on a conditional multinomial goodness-of-fit test. To capture as many effect-sharing patterns as possible, we choose to set the number of principal components in the ED step to be large. 
```{r warning=FALSE, message=FALSE}
##  identify the subset of differentially expressed genes and initialize the data-driven prior covariances
res.pca <- pois_cov_init(data, s, npc=26)
subset <- res.pca$subset
data.ed <- data[subset,]

## the number of differentially expressed genes identified by the multinomial goodness-of-fit test
dim(data.ed)

## load in the ED results
fit.ed <- readRDS("output/realdata/pois_mash_ed_maxiter2000.Rds")

## running time per iteration in ED
fit.ed$runtime/length(fit.ed$ELBO)

## estimated weights for different candidate data-driven prior covariances
names(fit.ed$pi) <- c(names(res.pca$Ulist), names(res.pca$ulist), "u_0")
fit.ed$pi

## look at the ELBO
plot(fit.ed$ELBO, xlab = "niter", ylab = "elbo", main="fitting ED step")

```



### Estimate of data-driven covariances from ED step
```{r warning=F}
library(RColorBrewer)

cols.all <- c("blue", brewer.pal(n=5,name="Greens")[3:5], "gray", brewer.pal(n=9,name="PuRd"), 
              brewer.pal(n=9,name="Reds"), brewer.pal(n=9,name="Oranges"),
              brewer.pal(n=9,name="Purples"), brewer.pal(n=9,name="YlOrRd"))

cols.trt <- cols.all[5:50]

par(mfrow=c(2,2))

pi.g <- fit.ed$pi[-c(1:2)]
ulist <- fit.ed$ulist
names(ulist) <- names(fit.ed$pi[-c(1:2)])

### plot the data-driven u_g's with estimated weights larger than or equal to 2%
ulist <- ulist[pi.g >= 2e-2]
pi.g <- pi.g[pi.g >= 2e-2]

for (k in 1:length(ulist)){
  v <- ulist[[k]]
  
  barplot(v/v[which.max(abs(v))], names = names(s), cex.names = 0.4, las = 2, 
          main = paste0(names(ulist)[k], ": weight ", round(pi.g[k],3)), col = cols.trt)
  
}

```



### Running poisson mash
We then fit the poisson mash model to all the genes in $X$ which have at least 20 total read counts in all conditions. There are 10,733 such genes.
```{r warning=FALSE, message=FALSE}
## remove genes that have fewer than 20 total read counts in all 46 conditions
data <- data[rowSums(data) >= 20,]
dim(data)

## load in the poisson mash fit results
res <- readRDS("output/realdata/pois_mash_fit.Rds")
res$runtime

## look at ELBO
plot(res$ELBO, xlab = "niter", ylab = "elbo", main="Fitting poisson mash with random effects")

## look at the estimated weights of prior covariance matrices
library(pheatmap)
library(gridExtra)

wlist <- c(1e-4, 4e-4, 1e-3, 4e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32, 0.64, 1, 2, 4, 8, 16)

pi.mat <- matrix(res$pi, ncol=length(wlist), byrow=TRUE)

rownames(pi.mat) <- c("tPCA", "Emp_cov", paste0("e_", 1:46), paste0("PC_", 1:26), "u_0")
colnames(pi.mat) <- paste0("w=", wlist)
pheatmap(pi.mat, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=6, main="Estimated weights of prior covariances in poisson mash fit")

```



### Comparison to fixed effects-based mash
We calculate posterior summaries of $\delta_{jr} = \beta_{jr} - \sum_{r=1}^R \beta_{jr}/R$ from the poisson mash fit, i.e., the condition-specific deviation from the mean across all conditions, and compare to the results from fixed effects-based mash. In particular, we fit a linear fixed effects model separately to each gene $j$, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the predictor, to obtain $\hat{b}_{jr}$ and $\hat{s}_{jr}$ for each gene $j$ and condition $r$. When fitting this model, we assume that the Gaussian observation noise is constant over conditions for each gene $j$. We then feed $\hat{b}_{jr}$ and $\hat{s}_{jr}$ to mash without common baseline. We consider only the 7,429 genes that have reads in at least $46*5=230$ cells.
```{r warning=FALSE, message=FALSE}
## load in the posterior summaries from poisson mash fit
post <- readRDS("output/realdata/pois_mash_posterior.Rds")

## load in the mash fit based on bhat and shat obtained by fitting fixed effects model
mash.fixed <- readRDS("output/realdata/mash_fit.rds")

## the number of significant genes identified by poisson mash (in 10,733 genes tested by poisson mash)
idx.pois.mash <- which(apply(post$lfsr, 1, min) < 0.05)
length(idx.pois.mash)

## the number of significant genes identified by fixed effects-based mash (in 7,429 genes tested by mash)
idx.mash.fixed <- mashr::get_significant_results(mash.fixed)
length(idx.mash.fixed)

## the number of genes that are identified by both poisson mash and fixed effects-based mash
sum(names(idx.mash.fixed) %in% names(idx.pois.mash))

## the number of genes that are identified only by poisson mash (in 7,429 genes tested by both)
pois_mash_only <- idx.pois.mash[!(names(idx.pois.mash) %in% names(idx.mash.fixed)) & (names(idx.pois.mash) %in% rownames(mash.fixed$result$lfsr))]
length(pois_mash_only)

## the number of genes that are identified only by mash 
mash_fixed_only <- idx.mash.fixed[!(names(idx.mash.fixed) %in% names(idx.pois.mash))]
length(mash_fixed_only)

```



### Comparison to mixed effects-based mash
We then compare results from the poisson mash to the results from mixed effects-based mash. In particular, we fit a linear mixed effects model separately to each gene $j$, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the design matrix for both fixed and random effects, to obtain $\hat{b}_{jr}$ and $\hat{s}_{jr}$ for each gene $j$ and condition $r$. We use the dream function in the R package variancePartition to fit this mixed effects model (https://www.bioconductor.org/packages/release/bioc/html/variancePartition.html). We then feed $\hat{b}_{jr}$ and $\hat{s}_{jr}$ to mash without common baseline. Again, we consider only the 7,429 genes that have reads in at least $46*5=230$ cells. 
```{r warning=FALSE, message=FALSE}
## load in the mash fit based on bhat and shat obtained by fitting mixed effects model
mash.mixed <- readRDS("output/realdata/mash_dream_fit.Rds")

## the number of significant genes identified by mixed effects-based mash (in 7,429 genes tested by mash)
idx.mash.mixed <- mashr::get_significant_results(mash.mixed)
length(idx.mash.mixed)

## the number of genes that are identified by both poisson mash and mixed effects-based mash
sum(names(idx.mash.mixed) %in% names(idx.pois.mash))

```