---
title: "Application of poisson mash and mash to simulated single cell data with signal (setting 3)"
author: "Yusha Liu"
header-includes:
date: "2020-8-3"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply poisson mash and mash to simulated single cell data with differential expression across conditions for a subset of genes using thinning. 

We start with a subset of the single cell cytokines data, which include 1) 6,038 cells from 21 conditions from the second batch, and 2) 7,492 genes that have at least 100 total counts over these 21 conditions. We first fit poisson mash with random effects to this dataset, and use the covariance matrices (both canonical and data-driven ones) learned from this poisson mash fit to simulate signals later. There are 3 data-driven covariances (each assigned $30\%$ weight) and 2 canonical covariances modeling condition-specific effects (each assigned $5\%$ weight) in total.

Starting with the real dataset, we randomly permute the cells with respect to the condition labels independently for each gene, so that there does not exist association between gene expression levels and conditions. Then we use thinning (https://cran.r-project.org/web/packages/seqgendiff/index.html) to add a known amount of signal to the randomly permuted data. Before thinning, we remove genes that have fewer than 200 total counts over all conditions, so there are 5,859 genes left. Then for each replicate dataset, we add signals to a randomly chosen subset of 500 genes. The signals are added in terms of differential effects across conditions on the log scale, which are simulated from a multivariate normal distribution with zero mean and covariance drawn randomly from one of the covariances learned from the poisson mash fit of the original dataset. The signals are further scaled appropriately to have realistic effect sizes (in terms of $\max_r \mid \beta_{jr}\mid$).

We simulate 20 single cell datasets in total.


### Plot ROC for detecting non-null units
See [plots](https://github.com/YushaLiu/poisson_mash/blob/master/output/non_null_units_setting3.pdf). \


### Plot ROC for detecting non-null unit-condition pairs
See [plots](https://github.com/YushaLiu/poisson_mash/blob/master/output/non_null_pairs_setting3.pdf). \


### Summary of results at a threshold of lfsr=0.05 for each replicate
```{r warning=FALSE, message=FALSE}
tpr.pois <- rep(NA, 20)
fpr.pois <- rep(NA, 20)
fdr.pois <- rep(NA, 20)
tpr.mash <- rep(NA, 20)
fpr.mash <- rep(NA, 20)
fdr.mash <- rep(NA, 20)
tpr.pair.pois <- rep(NA, 20)
fpr.pair.pois <- rep(NA, 20)
fsr.pair.pois <- rep(NA, 20)
tpr.pair.mash <- rep(NA, 20)
fpr.pair.mash <- rep(NA, 20)
fsr.pair.mash <- rep(NA, 20)

for(iter in 1:20){
  # load in the ground truth
  load(paste0("output/non_null_sim_setting3/scdata", iter, ".RData"))
  J <- nrow(beta)
  R <- ncol(beta)
  C <- diag(R) - (1/R)
  beta.contrast <- beta %*% C
  
  # poisson mash
  post <- readRDS(paste0("output/non_null_sim_setting3/pois_mash_posterior_rep_", iter, ".Rds"))

  # calculate TPR, FPR and FDR for detecting non-null units
  idx.pois <- which(apply(post$lfsr, 1, min) < 0.05)
  tpr.pois[iter] <- sum(names(idx.pois) %in% names(non.null.idx))/length(non.null.idx)
  fpr.pois[iter] <- sum(!names(idx.pois) %in% names(non.null.idx))/(J-length(non.null.idx))
  fdr.pois[iter] <- sum(!names(idx.pois) %in% names(non.null.idx))/length(idx.pois)
    
  # calculate TPR, FPR and FSR for detecting non-null unit-condition pairs
  sign.pair <- post$PosteriorMean * beta.contrast
  tpr.pair.pois[iter] <- sum(post$lfsr < 0.05 & sign.pair > 0)/sum(beta.contrast!=0)
  fpr.pair.pois[iter] <- sum(post$lfsr < 0.05 & beta.contrast==0)/sum(beta.contrast==0)
  fsr.pair.pois[iter] <- 1 - sum(post$lfsr < 0.05 & sign.pair > 0)/sum(post$lfsr < 0.05)
  
  
  # mash
  mash.fit <- readRDS(paste0("output/non_null_sim_setting3/mash_fit_rep_", iter, ".Rds"))
  
  # calculate TPR, FPR and FDR for detecting non-null units
  idx.mash <- mashr::get_significant_results(mash.fit, 0.05)
  tpr.mash[iter] <- sum(names(idx.mash) %in% names(non.null.idx))/length(non.null.idx)
  fpr.mash[iter] <- sum(!names(idx.mash) %in% names(non.null.idx))/(J-length(non.null.idx))
  fdr.mash[iter] <- sum(!names(idx.mash) %in% names(non.null.idx))/length(idx.mash)
    
  # calculate TPR, FPR and FSR for detecting non-null unit-condition pairs
  lfsr.pair <- ashr::get_lfsr(mash.fit)
  mean.pair <- ashr::get_pm(mash.fit)
  sign.pair <- mean.pair * beta.contrast
  tpr.pair.mash[iter] <- sum(lfsr.pair < 0.05 & sign.pair > 0)/sum(beta.contrast!=0)
  fpr.pair.mash[iter] <- sum(lfsr.pair < 0.05 & beta.contrast==0)/sum(beta.contrast==0)
  fsr.pair.mash[iter] <- 1 - sum(lfsr.pair < 0.05 & sign.pair > 0)/sum(lfsr.pair < 0.05) 
  
  rm(X, beta, non.null.idx)
}

```


### TPR, FPR and FDR for detecting non-null units at lfsr=0.05 for each replicate
```{r warning=FALSE, message=FALSE}
tpr.pois

fpr.pois

fdr.pois

tpr.mash

fpr.mash

fdr.mash

```


### TPR, FPR and FSR for detecting non-null unit-condition pairs at lfsr=0.05 for each replicate
```{r warning=FALSE, message=FALSE}
tpr.pair.pois

fpr.pair.pois

fsr.pair.pois

tpr.pair.mash

fpr.pair.mash

fsr.pair.mash

```