---
title: "Examples of genes that are identified by poisson mash with random effects only"
author: "Yusha Liu"
header-includes:
date: "2020-7-20"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Summary
We apply both poisson mash and mash to CD4 T cells of the single cell cytokines data. The four conditions from the first batch (Ctrl_1, IFNa, IFNb, TNF) are removed from analysis for now. 

```{r warning=FALSE, message=FALSE}
### load in functions
source("code/util_random_effects_v2.R")

### load in data
load("data/data_jr.RData")
data <- data.jr
s <- s.r/8
rm(data.jr, s.r)
data <- data[,5:50]
s <- s[5:50]

### remove genes that have fewer than 20 total counts across all 46 treatments
data <- data[rowSums(data) >= 20,]

### color code for treatments
library(RColorBrewer)
cols.all <- c("blue", brewer.pal(n=5,name="Greens")[3:5], "gray", brewer.pal(n=9,name="PuRd"), 
              brewer.pal(n=9,name="Reds"), brewer.pal(n=9,name="Oranges"),
              brewer.pal(n=9,name="Purples"), brewer.pal(n=9,name="YlOrRd"))
cols.trt <- cols.all[5:50]

## load in the ED results
fit.ed <- readRDS("output/ed_second_batch_v2_maxiter2000.Rds")

### load in posterior summary of pois mash
post <- readRDS("output/pois_mash_contrast_posterior_second_batch_v2.Rds")

### load in posterior summary of mash
mash.fit <- readRDS("output/mash_fit_second_batch.rds")

```


Now we present some examples of genes that are identified to be differentially expressed across treatments only by poisson mash with random effects, but also get tested in mash. 

Examples for which the first eigenvector in data-driven U1 has the largest weight:
```{r warning=FALSE, message=FALSE}
eig.tPCA <- eigen(fit.ed$Ulist[[1]])
pve.tPCA <- eig.tPCA$values/sum(eig.tPCA$values)
v <- eig.tPCA$vectors[,1]
barplot(v/v[which.max(abs(v))], names = names(s), cex.names = 0.6,
          las = 2, main = paste0(names(fit.ed$Ulist)[1], ": Eigenvector ", 1, " (pve = ", round(pve.tPCA[1],3), ")"), col = cols.trt)

create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Clk1", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Ivns1abp", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Fam78a", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Ccnl1", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Rprd2", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Ttyh3", cols.trt=cols.trt)

```


Examples for which the second eigenvector in data-driven U1 has the largest weight:
```{r warning=FALSE, message=FALSE}
v <- eig.tPCA$vectors[,2]
barplot(v/v[which.max(abs(v))], names = names(s), cex.names = 0.6,
          las = 2, main = paste0(names(fit.ed$Ulist)[1], ": Eigenvector ", 2, " (pve = ", round(pve.tPCA[2],3), ")"), col = cols.trt)

create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Sept6", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Sept7", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Psap", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Srrm2", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Jund", cols.trt=cols.trt)

```


Examples for which a condition-specific effect has the largest weight:
```{r warning=FALSE, message=FALSE}
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Tsc22d4", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Vgll4", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Rexo2", cols.trt=cols.trt)
create.plot(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename="Myc", cols.trt=cols.trt)

```