---
title: "Examples of genes that are identified by both poisson mash with random effects and mash"
author: "Yusha Liu"
header-includes:
date: "2020-8-3"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Summary
We apply both poisson mash and mash to CD4 T cells of the single cell cytokines data. The four conditions from the first batch (Ctrl_1, IFNa, IFNb, TNF) are removed from analysis for now. Now we present some examples of genes that are identified to be differentially expressed across treatments by both poisson mash with random effects and mash. 
```{r warning=FALSE, message=FALSE}
### load in functions
source("code/util_random_effects_v2.R")

### load in data
load("data/data_jr.RData")
data <- data.jr
s <- s.r/8
rm(data.jr, s.r)
data <- data[,5:50]
s <- s[5:50]

### remove genes that have fewer than 20 total counts across all 46 treatments
data <- data[rowSums(data) >= 20,]

### color code for treatments
library(RColorBrewer)
cols.all <- c("blue", brewer.pal(n=5,name="Greens")[3:5], "gray", brewer.pal(n=9,name="PuRd"), 
              brewer.pal(n=9,name="Reds"), brewer.pal(n=9,name="Oranges"),
              brewer.pal(n=9,name="Purples"), brewer.pal(n=9,name="YlOrRd"))
cols.trt <- cols.all[5:50]

### load in posterior summary of pois mash
post <- readRDS("output/pois_mash_contrast_posterior_second_batch_v2.Rds")
lfsr.pois.mash <- post$lfsr
idx.pois.mash <- which(apply(lfsr.pois.mash, 1, min) < 0.01)

### load in posterior summary of mash
mash.fit <- readRDS("output/mash_fit_second_batch.rds")
lfsr.mash <- mash.fit$result$lfsr
idx.mash <- mashr::get_significant_results(mash.fit, 0.01)

### genes that are identified to be differentially expressed by both poisson mash and mash
idx.both <- idx.pois.mash[names(idx.pois.mash) %in% names(idx.mash)]

### randomly select 10 genes to plot
set.seed(500)
genes.to.plot <- names(idx.both)[sample(1:length(idx.both), 10, replace = FALSE)]

```


```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[1], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[2], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[3], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[4], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[5], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[6], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[7], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[8], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[9], cols.trt=cols.trt)
```

```{r warning=FALSE, message=FALSE}
create.plot.v2(data=data, s=s, res.pois.mash=post, res.mash=mash.fit$result, genename=genes.to.plot[10], cols.trt=cols.trt)
```