---
title: "Application of poisson mash and mash to simulated single cell data (Ly6C-) under the null"
author: "Yusha Liu"
header-includes:
date: "2020-9-16"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply poisson mash and mash to single cell data simulated under the null. More specifically, we start with a subset of the single cell cytokines data, which include 1) 1,879 cells from 25 conditions from the second batch, and 2) 6,794 genes that have at least 100 total counts over these 25 conditions. Then we randomly permute the cells with respect to the condition labels so that there no longer exists association between gene expression levels and conditions. 

We consider three different methods to simulate null data based on the real dataset: 

1). apply the same permutation to all the genes, so the correlation among genes is preserved on the permuted dataset; 

2). apply independent and different permutation to each gene; 

3). apply 1) followed by gene-wise addition of random effects via thinning to the randomly permuted data.

For each of the three simulation methods described above, we simulate 5 single cell datasets. 

For method 1), the number of false positives are $98, 78, 160, 92, 160$ for poisson mash, and $2, 2, 2, 2, 4$ for fixed effects-based mash.

For method 2), the number of false positives are $11, 12, 14, 12, 16$ for poisson mash, and $3, 2, 4, 2, 0$ for fixed effects-based mash.

For method 3), the number of false positives are $30, 25, 25, 100, 44$ for poisson mash, $83, 94, 90, 98, 90$ for fixed effects-based mash, and $7, 3, 6, 10, 7$ for mixed effects-based mash. 


## Exploration of false positives found by poisson mash using method 1)
```{r warning=FALSE, message=FALSE}
library(Matrix)
library(pheatmap)
library(gridExtra)

### for one replicate, load in the raw count data at the cell level, and the poisson mash fit
data <- readRDS("output/Ly6C-/sim_null/raw_data3.Rds")
fit <- readRDS("output/Ly6C-/sim_null/ed_rep_3.Rds")
res <- readRDS("output/Ly6C-/sim_null/pois_mash_fit_rep_3.Rds")
post <- readRDS("output/Ly6C-/sim_null/pois_mash_posterior_rep_3.Rds")

### look at the posterior weights of the genes identified (false positives) 
genes <- which(apply(post$lfsr, 1, min) < 0.05)
wlist <- c(1e-3, 2.5e-3, 5e-3, 1e-2, 2e-2, 4e-2, 8e-2, 0.16, 0.32, 0.64, 1, 2, 4, 8, 16, 25, 36, 50)
Ulist <- c(names(fit$Ulist), paste0("e_", 1:25), names(fit$ulist))
zeta.max <- rep(NA, length(genes))

for(j in 1:length(genes)){
  zeta <- matrix(res$Ez[genes[j],], ncol=length(wlist), byrow = TRUE)
  zeta.max[j] <- Ulist[which.max(rowSums(zeta))]
}

table(zeta.max)


### identified genes for which the data-driven tPCA has the largest weight 
gene.list <- genes[zeta.max=="tPCA"]

### for this subset of genes, compute the gene-wise correlation across all cells
data.sub <- as.matrix(data[gene.list,])
corr.sub <- cor(t(data.sub))

pheatmap(corr.sub, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=4, main="Gene-wise correlation for the identified genes \n (largest weight: tPCA)")


### for a randomly chosen subset of genes of the same size, compute the gene-wise correlation across all cells
set.seed(100)
data.tmp <- as.matrix(data[sample(1:nrow(data), length(gene.list), replace=FALSE),])
corr.tmp <- cor(t(as.matrix(data.tmp)))

pheatmap(corr.tmp, cluster_rows=FALSE, cluster_cols=FALSE, fontsize_row=4, fontsize_col=4, main="Gene-wise correlation for a randomly chosen set of genes")

```



## Conclusions
1. For poisson mash, the number of false positives is greatly reduced if we apply method 2) rather than method 1). The explorations above suggest that the false positives found by poisson mash when using method 1) could be attributed to the strong gene-wise correlation present in the original data, which are preserved on the permuted dataset.  

2. When random effects are not present, the number of false positives found by fixed effects-based mash is always smaller than that of poisson mash. Since the number of false positives is already negligibly small for fixed effects-based mash, we do not consider mixed effects-based mash in this setting.

3. When random effects are present, the number of false positives found by fixed effects-based mash tends to be larger than that of poisson mash. Mixed effects-based mash always finds the smallest number of false positives. 