---
title: "Application of poisson mash and mash to simulated single cell data under the null"
author: "Yusha Liu"
header-includes:
date: "2020-10-1"
output: workflowr::wflow_html
editor_options:
  chunk_output_type: console
---

## Simulation procedure
We apply poisson mash and mash to single cell data simulated under the null. More specifically, we start with a subset of the single cell cytokines data, which include 1) 9,902 cells from 25 conditions, and 2) 8,568 genes that have at least 100 total counts over these 25 conditions. Then we consider two ways to simulate null data based on the real dataset: 

1). randomly permute the cells with respect to the condition labels so that there no longer exists association between gene expression levels and conditions;

2). apply 1) followed by addition of gene-specific random effects via thinning to the randomly permuted data (https://cran.r-project.org/web/packages/seqgendiff/index.html). We choose $\phi_j$, which controls the gene-specific random effect size, from the set $\log_2(\exp(1))*\{0.01, 0.05, 0.1, 0.2\}$ with probability $\{0.4, 0.4, 0.1, 0.1\}$, so that it resembles the empirical distribution of $\phi_j$ we see from the poisson mash fit to the real data.   



## Description of competing approaches
Fixed effects-based mash: we fit a linear fixed effects model separately to each gene $j$, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the predictor, to obtain $\hat{b}_{jr}$ and $\hat{s}_{jr}$ for each gene $j$ and condition $r$. When fitting this model, we assume that the Gaussian observation noise is constant over conditions for each gene $j$. We then feed $\hat{b}_{jr}$ and $\hat{s}_{jr}$ to mash without common baseline. 

Mixed effects-based mash: we fit a linear mixed effects model separately to each gene $j$, using the single-cell level normalized gene expression data as the response and the condition assignment of each cell as the design matrix for both fixed and random effects, to obtain $\hat{b}_{jr}$ and $\hat{s}_{jr}$ for each gene $j$ and condition $r$. We use the dream function in the R package variancePartition to fit this mixed effects model (https://www.bioconductor.org/packages/release/bioc/html/variancePartition.html). We then feed $\hat{b}_{jr}$ and $\hat{s}_{jr}$ to mash without common baseline.



## Results
For each of the two scenarios described above, we simulate 5 single cell datasets. 

For scenario 1), the number of false positives are $46, 15, 28, 22, 56$ for poisson mash, and $6, 3, 12, 5, 6$ for fixed effects-based mash.

For scenario 2), the number of false positives are $20, 9, 9, 10, 8$ for poisson mash, $338, 348, 329, 310, 310$ for fixed effects-based mash, and $36, 34, 29, 30, 35$ for mixed effects-based mash. 



## Conclusions
\noindent 1. When random effects are not present, the number of false positives found by fixed effects-based mash is always smaller than that of poisson mash. Since the number of false positives is already negligibly small for fixed effects-based mash, we do not consider mixed effects-based mash in this setting.

\noindent 2. When random effects are present, the number of false positives found by fixed effects-based mash is much larger than that of poisson mash, which finds the smallest number of false positives among the three methods (poisson mash, fixed effects-based mash, and mixed effects-based mash).